nighthawk/source/common/sequencer_impl.cc

```diff
@@ -0,0 +1,145 @@
+#include "nighthawk/source/common/sequencer_impl.h"
+
+#include "common/common/assert.h"
+
+#include "nighthawk/common/exception.h"
+
+#include "nighthawk/source/common/platform_util_impl.h"
+
+using namespace std::chrono_literals;
+
+namespace Nighthawk {
+
+constexpr std::chrono::milliseconds SequencerImpl::EnvoyTimerMinResolution = 1ms;
+
+SequencerImpl::SequencerImpl(PlatformUtil& platform_util, Envoy::Event::Dispatcher& dispatcher,
+                             Envoy::TimeSource& time_source, RateLimiter& rate_limiter,
+                             SequencerTarget& target, std::chrono::microseconds duration,
+                             std::chrono::microseconds grace_timeout)
+    : target_(target), platform_util_(platform_util), dispatcher_(dispatcher),
+      time_source_(time_source), rate_limiter_(rate_limiter), duration_(duration),
+      grace_timeout_(grace_timeout), start_(time_source.monotonicTime().min()),
+      targets_initiated_(0), targets_completed_(0), running_(false), blocked_(false) {
+  ASSERT(target_ != nullptr, "No SequencerTarget");
+  periodic_timer_ = dispatcher_.createTimer([this]() { run(true); });
```

Comment:
Well, the periodic timer gets re-enabled by `run()`, and the adhoc timer does not. We're not *always* busy-looping, with long-running transactions the periodic timer is mostly driving this, and gives us a minimum of timely initiating new work. Tangentially related, I think that if we could somehow prioritize i/o events over timer events in libevent when we schedule a timer event to fire immediately, then I think we could always busy-loop and thus drop the periodic timer.

Reply:
My main concern is around complexity. Questions:
1. For the periodic timer, are we making a latency tradeoff? Is the worst outcome that the release of requests is a bit off, but our latency measurement once released is not affected?
2. Can you write up a detailed description of this interplay somewhere? It's kind of confusing to the reader what these are.
3. Would it make sense to rename `adhoc_timer_` as `wait_for_ratelimiter_timer_`?

Reply:
1. Timing is way more accurate

Reply:
One more thing I wanted to add:
Observationally, it looks like the spin loop has a very positive effect on some systems. The laptop I am commenting with right now is not tuned for latency and has a i7-8750H cpu. 
Mean latency reported by wrk2 in a 1000rps test is 299usec, whereas NH reports 115 usec.

Reply:
PS: Experimentally removing the spin loop, NH reports a mean of 336 usec.

Reply:
Yeah, this is pretty interesting. Maybe we want two modes of operation, one in which we always spin, and another in which we only spin when waiting for the rate limiter. I think "always spin" might be interesting for two reasons:
1. If the rate limiter is bursty at the libevent timer granularity, always spinning would allow us to achieve the finest granularity possible.
2. We would avoid having any CPU or OS level power management kick-in, by always maintaining the load generator in maximum C state.
This is of course, completely counter-intuitive from an energy efficiency or shared compute environment perspective, but I think it makes a lot of sense for load generation and latency measurement.

> Perhaps rename `adhoc_timer` to `wait_for_next_timer_`?

Sure.

Reply:
Nice, `spin_timer_` makes sense, done. 

Regarding always spinning: I actually tried that a little while ago. When doing that we unfortunately won't get any I/O events. So getting that to work may need a change in libevent or its configuration when we set it up. I do remember code mentioning priorities in the libevent code base, and in this particular perhaps prioritizing I/O events over timers would be what we need.

Reply:
That's interesting. Do we lose I/O events when spinning for rate limiter admittance as well? In any case, I'm sold on what you have here as a starting point, what I think we need to do to get this merged is to transfer this deep dive that we've been going through into a bunch of comments in the `.h` and `.cc`; most readers won't have any of this intuition! :)